File: app/src/androidTest/java/de/westnordost/streetcomplete/data/osmnotes/CreateNoteDaoTest.java
package de.westnordost.streetcomplete.data.osmnotes;

import java.util.ArrayList;

import de.westnordost.streetcomplete.data.ApplicationDbTestCase;
import de.westnordost.osmapi.map.data.BoundingBox;
import de.westnordost.osmapi.map.data.Element;
import de.westnordost.osmapi.map.data.OsmLatLon;

public class CreateNoteDaoTest extends ApplicationDbTestCase
{
	private CreateNoteDao dao;

	@Override public void setUp() throws Exception
	{
		super.setUp();
		dao = new CreateNoteDao(dbHelper, serializer);
	}

	public void testAddGetDelete()
	{
		CreateNote note = new CreateNote();
		note.text = "text";
		note.questTitle = "title";
		note.position = new OsmLatLon(3,5);
		note.elementId = 123L;
		note.elementType = Element.Type.NODE;
		note.imagePaths = new ArrayList<>();
		note.imagePaths.add("hullo");
		note.imagePaths.add("hey");

		assertTrue(dao.add(note));
		CreateNote dbNote = dao.get(note.id);

		assertEquals(note.text, dbNote.text);
		assertEquals(note.questTitle, dbNote.questTitle);
		assertEquals(note.id, dbNote.id);
		assertEquals(note.position, dbNote.position);
		assertEquals(note.elementType, dbNote.elementType);
		assertEquals(note.elementId, dbNote.elementId);
		assertEquals(note.imagePaths, dbNote.imagePaths);

		assertTrue(dao.delete(note.id));

		assertNull(dao.get(note.id));
	}

	public void testNullable()
	{
		CreateNote note = new CreateNote();
		note.text = "text";
		note.position = new OsmLatLon(3,5);

		assertTrue(dao.add(note));
		CreateNote dbNote = dao.get(note.id);

		assertNull(dbNote.elementType);
		assertNull(dbNote.elementId);
		assertNull(dbNote.questTitle);
		assertNull(dbNote.imagePaths);
	}

	public void testGetAll()
	{
		CreateNote note1 = new CreateNote();
		note1.text = "this is in";
		note1.position = new OsmLatLon(0.5,0.5);
		dao.add(note1);

		CreateNote note2 = new CreateNote();
		note2.text = "this is out";
		note2.position = new OsmLatLon(-0.5,0.5);
		dao.add(note2);

		assertEquals(1,dao.getAll(new BoundingBox(0,0,1,1)).size());

		assertEquals(2,dao.getAll(null).size());
	}


	public void testCountNone()
	{
		assertEquals(0, dao.getCount());
	}

	public void testCountOne()
	{
		CreateNote note1 = new CreateNote();
		note1.position = new OsmLatLon(0.5,0.5);
		note1.text = "joho";
		dao.add(note1);

		assertEquals(1, dao.getCount());
	}

	public void testCountSeveral()
	{
		CreateNote note1 = new CreateNote();
		note1.position = new OsmLatLon(0.5,0.5);
		note1.text = "joho";
		dao.add(note1);

		CreateNote note2 = new CreateNote();
		note2.position = new OsmLatLon(0.1,0.5);
		note2.text = "hey";
		dao.add(note2);

		assertEquals(2, dao.getCount());
	}
}

File: app/src/androidTest/java/de/westnordost/streetcomplete/data/osmnotes/CreateNoteDaoTest.kt
package de.westnordost.streetcomplete.data.osmnotes

import org.junit.Before
import org.junit.Test

import de.westnordost.streetcomplete.data.ApplicationDbTestCase
import de.westnordost.osmapi.map.data.BoundingBox
import de.westnordost.osmapi.map.data.Element
import de.westnordost.osmapi.map.data.OsmLatLon
import de.westnordost.streetcomplete.data.osm.ElementKey

import org.junit.Assert.*

class CreateNoteDaoTest : ApplicationDbTestCase() {
    private lateinit var dao: CreateNoteDao

    @Before fun createDao() {
        dao = CreateNoteDao(dbHelper, CreateNoteMapping(serializer))
    }

    @Test fun getButNothingIsThere() {
        assertNull(dao.get(1L))
    }

    @Test fun getAllButNothingIsThere() {
        assertEquals(listOf<CreateNote>(), dao.getAll())
    }

    @Test fun addAndGet() {
        val note = CreateNote(null, "text", OsmLatLon(3.0, 5.0), "title",
            ElementKey(Element.Type.NODE, 132L), arrayListOf("hello","hey"))

        assertTrue(dao.add(note))
        val dbNote = dao.get(note.id!!)!!

        assertEquals(note, dbNote)

        assertTrue(dao.delete(note.id!!))

        assertNull(dao.get(note.id!!))
    }

    @Test fun delete() {
        val note = CreateNote(null, "text", OsmLatLon(3.0, 5.0))

        assertTrue(dao.add(note))
        assertTrue(dao.delete(note.id!!))
        assertNull(dao.get(note.id!!))
    }

    @Test fun addAndGetNullableFields() {
        val note = CreateNote(null, "text", OsmLatLon(3.0, 5.0))

        assertTrue(dao.add(note))
        val dbNote = dao.get(note.id!!)!!

        assertNull(dbNote.elementKey)
        assertNull(dbNote.questTitle)
        assertNull(dbNote.imagePaths)
    }

    @Test fun getAll() {
        dao.add(CreateNote(null, "this is in", OsmLatLon(0.5, 0.5)))
        dao.add(CreateNote(null, "this is out", OsmLatLon(-0.5, 0.5)))

        assertEquals(1, dao.getAll(BoundingBox(0.0, 0.0, 1.0, 1.0)).size)
        assertEquals(2, dao.getAll().size)
    }

    @Test fun getCount0() {
        assertEquals(0, dao.getCount())
    }

    @Test fun getCount1() {
        dao.add(CreateNote(null, "joho", OsmLatLon(0.5, 0.5)))
        assertEquals(1, dao.getCount())
    }

    @Test fun getCount2() {
        dao.add(CreateNote(null, "joho", OsmLatLon(0.5, 0.5)))
        dao.add(CreateNote(null, "joho", OsmLatLon(0.1, 0.5)))

        assertEquals(2, dao.getCount())
    }
}
